flang/lib/Lower/OpenMP.cpp

```diff
@@ -715,6 +715,188 @@ static void genOMP(Fortran::lower::AbstractConverter &converter,
                                 &wsLoopOpClauseList, iv);
 }
 
+static mlir::StringAttr handleOmpAtomicWriteMemoryOrderClause(
+    Fortran::lower::AbstractConverter &converter,
+    const Fortran::parser::OmpAtomicClause &clause) {
+  mlir::StringAttr memory_order;
+  auto &firOpBuilder = converter.getFirOpBuilder();
+  if (auto ompMemoryOrderClause =
+          std::get_if<Fortran::parser::OmpMemoryOrderClause>(&clause.u)) {
+    // 'acq_rel' and 'acquire' are not valid memory order clauses under atomic
+    // write
+    if (std::get_if<Fortran::parser::OmpClause::Release>(
+            &ompMemoryOrderClause->v.u)) {
+      memory_order =
+          firOpBuilder.getStringAttr(omp::stringifyClauseMemoryOrderKind(
+              omp::ClauseMemoryOrderKind::release));
+    } else if (std::get_if<Fortran::parser::OmpClause::Relaxed>(
+                   &ompMemoryOrderClause->v.u)) {
+      memory_order =
+          firOpBuilder.getStringAttr(omp::stringifyClauseMemoryOrderKind(
+              omp::ClauseMemoryOrderKind::relaxed));
+    } else if (std::get_if<Fortran::parser::OmpClause::SeqCst>(
+                   &ompMemoryOrderClause->v.u)) {
+      memory_order =
+          firOpBuilder.getStringAttr(omp::stringifyClauseMemoryOrderKind(
+              omp::ClauseMemoryOrderKind::seq_cst));
+    }
+  }
+  return memory_order;
+}
+
+static mlir::StringAttr handleOmpAtomicReadMemoryOrderClause(
```

Comment:
This is still kind of redundant. With atomic update and capture constructs, you would need two more functions. Can you merge this and `handleOmpAtomicWriteMemoryOrderClause` and rename it as `getOmpAtomicWriteMemoryOrderAttr`? I think we can let semantic checks handle which memory order clause can be in which construct.